signer/xpi/xpi.go

```diff
@@ -143,60 +148,121 @@ func (s *PKCS7Signer) Config() signer.Configuration {
 	}
 }
 
-// SignFile takes an unsigned zipped XPI file and returned a signed XPI file
-func (s *PKCS7Signer) SignFile(input []byte, options interface{}) (signer.SignedFile, error) {
+// SignFile takes an unsigned zipped XPI file and returns a signed XPI file
+func (s *XPISigner) SignFile(input []byte, options interface{}) (signedFile signer.SignedFile, err error) {
 	var (
-		signedFile []byte
+		pkcs7Manifest []byte
+		manifest      []byte
+		metas         = []Metafile{}
+		opt           Options
+		coseSig       []byte
+		coseSigAlgs   []*cose.Algorithm
 	)
-	manifest, sigfile, err := makeJARManifests(input)
+
+	opt, err = GetOptions(options)
 	if err != nil {
-		return nil, errors.Wrap(err, "xpi: cannot make JAR manifests from XPI")
+		return nil, errors.Wrap(err, "xpi: cannot get options")
 	}
-	p7sig, err := s.signData(sigfile, options)
+	coseSigAlgs, err = opt.Algorithms()
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: error parsing COSE Algorithms from options")
+	}
+
+	manifest, err = makeJARManifest(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
+	}
+
+	if len(coseSigAlgs) < 1 {
+		pkcs7Manifest = manifest
+	} else {
+		cn, err := opt.CN(s)
+		if err != nil {
+			return nil, err
+		}
+		coseSig, err = s.issueCOSESignature(cn, manifest, coseSigAlgs)
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: error signing cose message")
+		}
+
+		// add the cose files to the metafiles we'll add to the XPI
+		coseMetaFiles := []Metafile{
+			{"META-INF/cose.manifest", manifest},
+			{"META-INF/cose.sig", coseSig},
+		}
+		metas = append(metas, coseMetaFiles...)
+
+		// add entries for the cose files to the manifest as cose.manifest and cose.sig
+		pkcs7Manifest, err = makePKCS7Manifest(input, metas)
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: error making PKCS7 manifest")
+		}
+	}
+
+	sigfile, err := makeJARSignature(pkcs7Manifest)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest signature from XPI")
+	}
+
+	opt.COSEAlgorithms = []string{}
+	p7sig, err := s.signData(sigfile, opt)
```

Comment:
should rename `s.signData` to `s.signDataWithPKCS7` to be explicit that you're not issuing cose signatures there.